src/puppetlabs/trapperkeeper/authorization/acl.clj

```diff
@@ -56,123 +82,165 @@
   "Checks that name matches the given ace"
-  [{:keys [pattern match]} :- Entry
+  [{:keys [value match] :as ace} :- ACE
    to-match :- schema/Str]
   (let [match-split-domain (split-domain to-match)]
     (if (= :regex match)
-      (boolean (re-find (re-pattern pattern) to-match))
-      (every? (fn [[a b]] (= a b)) (map vector pattern match-split-domain)))))
+      (boolean (re-find (re-pattern value) to-match))
+      (every? (fn [[a b]] (= a b)) (map vector value match-split-domain)))))
 
-(defn- substitute-backreference
+(schema/defn ^:private substitute-backreference :- String
   "substiture $1, $2... by the same index in the captures vector"
-  [in captures]
+  [in :- String
+   captures :- [String]]
   (clojure.string/replace in #"\$(\d+)" #(nth captures (- (read-string (second %)) 1))))
 
-(defn interpolate-backreference
+(schema/defn interpolate-backreference :- ACE
   "change all possible backreferences in ace patterns to values from the
   capture groups"
-  [ace captures]
-  (if (= (:match ace) :backreference)
-    (new-domain (ace :auth-type)
-                (clojure.string/join "." (map #(substitute-backreference % captures)
-                                              (reverse (ace :pattern)))))
+  [{:keys [match auth-type] :as ace} :- ACE
+   captures :- [String]]
+  (if (= match :backreference)
+    (new-domain auth-type
+                {:certname (clojure.string/join "." (map #(substitute-backreference % captures)
+                                                         (reverse (ace :value))))})
     ace))
 
+(schema/defn match-extensions? :- schema/Bool
+  "Returns true if the provided SSL extension map matches the configured ACE.
+  All of the keys in the ACE must appear in the extensions map and, if the value
+  for a key in the ACE is a list, at least one of the listed values must be set
+  in the incoming extensions map.
+
+  Note the behavior in the following scenario: If an ACE specifies
+  {:deny {:extensions {:pp_env 'test'
+                       :pp_image 'bad image'}}}
+
+  *ONLY* a request with both :pp_env set to 'test' and :pp_image set to 'bad
+  image' would be denied. If *any* request with :pp_role set to 'test' is to be
+  denied, it needs a standalone deny rule."
+  [ace :- ACE
+   extensions :- Extensions]
+  (let [match-key (fn [k]
+                    (let [ace-value (get (:value ace) k)
+                          ext-value (get extensions k false)]
+                      (if ext-value
+                        (if (sequential? ace-value)
+                          (some (partial = ext-value) ace-value)
+                          (= ace-value ext-value))
+                        false)))]
+    (every? match-key (keys (:value ace)))))
+
 (schema/defn match? :- schema/Bool
-  "Returns true if the given name matches the given ace"
-  [ace :- Entry
-   name :- schema/Str]
-  (match-name? ace name))
+  "Returns true if the given value matches the given ace"
+  [{:keys [match value] :as acl-ace} :- ACE
```

Comment:
Looks like `value` on this line and `incoming-ace` from the following line aren't used here.